mapper/thin_edge_json/src/json.rs

```diff
@@ -1,698 +1,736 @@
+        }
+    }
+
+    fn new_invalid_json_value(name: &str, json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonValue {
+            name: String::from(name),
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn new_invalid_json_time(json: &JsonValue) -> ThinEdgeJsonError {
+        ThinEdgeJsonError::InvalidThinEdgeJsonTime {
+            actual_type: ThinEdgeJsonError::json_type(&json).to_string(),
+        }
+    }
+
+    fn json_type(input: &JsonValue) -> &'static str {
+        match input {
+            JsonValue::String(_) | JsonValue::Short(_) => "a string",
+            JsonValue::Number(_) => "a number",
+            JsonValue::Object(_) => "an object",
+            JsonValue::Array(_) => "an array",
+            JsonValue::Boolean(_) => "a boolean",
+            JsonValue::Null => "null",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_str_with_invalid_timestamp() {
+        let input = r#"{
+            "time" : "2013-06-2217:03:14.000658767+02:00"
+        }"#;
+        let expected_error = r#"Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): "2013-06-2217:03:14.000658767+02:00": input contains invalid characters"#;
+        let output_err = ThinEdgeJson::from_str(input).unwrap_err();
+        assert_eq!(output_err.to_string(), expected_error);
+    }
+
+    #[test]
+    fn test_str_with_valid_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14+02:00",
+            "temperature" : 25
+        }"#;
+        let output = ThinEdgeJson::from_str(input).unwrap();
+        assert_eq!(
+            output.timestamp,
+            Some(
+                FixedOffset::east(2 * 3600)
+                    .ymd(2021, 4, 30)
+                    .and_hms(17, 3, 14)
+            )
+        );
+    }
+
+    #[test]
+    fn test_str_with_millisecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123+02:00",
+            "temperature" : 25
+        }"#;
+
+        let output = ThinEdgeJson::from_str(input).unwrap();
+        assert_eq!(
+            output.timestamp,
+            Some(
+                FixedOffset::east(2 * 3600)
+                    .ymd(2021, 4, 30)
+                    .and_hms_milli(17, 3, 14, 123)
+            )
+        );
+    }
+
+    #[test]
+    fn test_str_with_nanosecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123456789+02:00",
+            "temperature" : 25
+        }"#;
+
+        let output = ThinEdgeJson::from_str(input).unwrap();
+        assert_eq!(
+            output.timestamp,
+            Some(
+                FixedOffset::east(2 * 3600)
+                    .ymd(2021, 4, 30)
+                    .and_hms_nano(17, 3, 14, 123456789)
+            )
+        );
+    }
+
+    #[test]
+    fn test_has_a_timestamp() {
```

Comment:
You shouldn't prepend tests with `test_`, while run they will look like so: `test json::tests::test_has_a_timestamp ... ok`